NeoML/src/Dnn/ReferenceDnnFactory.cpp

```diff
@@ -0,0 +1,168 @@
+﻿/* Copyright © 2024 ABBYY
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/Dnn/Dnn.h>
+#ifdef NEOML_COMPACT // No optimizations in mobile assembly
+namespace NeoML {
+static void OptimizeDnn( CDnn& ) {}
+}
+#else  // !NEOML_COMPACT
+#include <NeoML/Dnn/DnnOptimization.h>
+#endif // !NEOML_COMPACT
+
+namespace NeoML {
+
+// Internal technical class
+struct CReferenceDnnInfo final {
+	CRandom Random; // Stores the dnn's own external random class inside this dnn class
+	CPtr<CReferenceDnnFactory> Owner; // The factory accounts the number of owned reference dnns
+
+	CReferenceDnnInfo( CRandom rand, CPtr<CReferenceDnnFactory> ptr ) :
+		Random( std::move( rand ) ), Owner( std::move( ptr ) ) {}
+};
+
+void CReferenceDnnInfoDeleter::operator()( CReferenceDnnInfo* info ) { delete info; }
+
+//---------------------------------------------------------------------------------------------------------------------
+
+CPtr<CReferenceDnnFactory> CReferenceDnnFactory::Create( IMathEngine& mathEngine,
+	CArchive& archive, int seed, bool optimizeDnn )
+{
+	CPtr<CReferenceDnnFactory> factory = new CReferenceDnnFactory( CRandom( seed ), mathEngine );
+	factory->serialize( archive, optimizeDnn );
+	return factory;
+}
+
+CPtr<CReferenceDnnFactory> CReferenceDnnFactory::Create( IMathEngine& mathEngine,
+	const CDnn& dnn, bool optimizeDnn )
+{
+	CPtr<CReferenceDnnFactory> factory = new CReferenceDnnFactory( dnn.Random(), mathEngine );
+	// Copy dnn using serialization to get the new dnn of necessary life time
+	CMemoryFile file;
+	CArchive archive( &file, CArchive::store );
+	const_cast<CDnn&>( dnn ).Serialize( archive );
+	archive.Close();
+	file.SeekToBegin();
+
+	archive.Open( &file, CArchive::load );
+	factory->serialize( archive, optimizeDnn );
+	return factory;
+}
+
+CPtr<CReferenceDnnFactory> CReferenceDnnFactory::Create( CDnn&& dnn, bool optimizeDnn )
+{
+	CPtr<CReferenceDnnFactory> factory = new CReferenceDnnFactory( dnn.Random(), dnn.GetMathEngine() );
+	auto& mathEngineCopy = dnn.GetMathEngine();
+	auto& randomCopy = dnn.Random();
+
+	// Allow to createReferenceDnn() this time
+	NeoAssert( !dnn.IsReferenceDnn() );
+	NeoAssert( factory->Origin->Dnn.IsReferenceDnn() );
+	swap( dnn.referenceDnnInfo, factory->Origin->Dnn.referenceDnnInfo );
```

Comment:
This is very confusing, we swap referenceDnnInfos, then pass the original dnn (`dnn`) to initializeReferenceDnn as original dnn, but inside that function there is an `assert( originalDnn.IsReferenceDnn() )`.
It's too hard to read, needs either more comments (with warning signs) or a better naming, or another refactoring.

Reply:
fixed